src/kernel.cpp

```diff
@@ -108,30 +325,74 @@ static bool SelectBlockFromCandidates(
     return fSelected;
 }
 
-/* NEW MODIFIER */
+unsigned int GetStakeModifierChecksum(const CBlockIndex* pindex)
+{
+    assert(pindex->pprev || pindex->GetBlockHash() == Params().HashGenesisBlock());
+    // Hash previous checksum with flags, hashProofOfStake and nStakeModifier
+    CDataStream ss(SER_GETHASH, 0);
+    if (pindex->pprev)
+        ss << pindex->pprev->nStakeModifierChecksum;
+    ss << pindex->nFlags << pindex->hashProofOfStake << pindex->nStakeModifier;
+    uint256 hashChecksum = Hash(ss.begin(), ss.end());
+    hashChecksum >>= (256 - 32);
+    return hashChecksum.Get64();
+}
 
-// Stake Modifier (hash modifier of proof-of-stake):
-// The purpose of stake modifier is to prevent a txout (coin) owner from
-// computing future proof-of-stake generated by this txout at the time
-// of transaction confirmation. To meet kernel protocol, the txout
-// must hash with a future stake modifier to generate the proof.
-uint256 ComputeStakeModifier(const CBlockIndex* pindexPrev, const uint256& kernel)
+bool GetOldStakeModifier(CStakeInput* stake, uint64_t& nStakeModifier)
```

Comment:
Have a minor point about this method. if my memory doesn't fail me, previously stakeInput was caching the modifier data. Now it will have to search it every time?